### DISCUSSION TITLE
Option to set amiberry GUI startup size?

### DISCUSSION THREAD

**Discussion by giantclambake** [BlitterStudio/amiberry#1079] [General]
The GUI by default starts at 800x600 window ~ wrt using amiberry as a desktop app on a modern, hires widescreen monitor, the GUI appears very small, and it would be better if the GUI started at 1024x768 imho (ease of use). Yes, I do realize one can drag/resize the GUI window, but it's not the same as having not to do that 8)

I couldn't find any way to achieve this and/or could this feature be added to amiberry (set via amiberry.conf options)?

**Comment by giantclambake** [comment]
I circled back to look at this, and now realize to do this means impacting upon scrollarea* & buttons (in left config pane of GUI), and other stuff so associated...ergo, not as straight-forward as I first imagined =) Mind you...I still think it's a good idea wrt the context above, however it's a real 'back burner' task that only becomes an issue on hires displays (I think mostly evident with HDMI connected teevee display)...

**Reply by midwan** [reply]
The GUI is restricted to 800x600 in order to support devices that use lower resolutions as well. There are cases of handhelds, smaller screens and even some people who use CRT monitors and real PAL output (720x568), so keeping those dimensions helps as the minimum supported mode.

Ideally it should adapt of course, but that's not possible (at least, not easily) with the current `guisan` minimal library used for the GUI. Perhaps if the whole GUI is rewritten in something more modern (like Qt), but that will also increase the dependencies and it's a complex task.

If anyone wants to assist with it, PRs are welcome of course. But for now, I think it will stay at that size.

**Comment by giantclambake** [comment]
Thanks, I sort of suspected compatibility with handhelds might be at the root of it =)

Like I say, one can start the GUI (800x600) and drag/resize the GUI window to your liking  --- it would be nice if amiberry had that Amiga workbench facility of being able to 'snapshot' the current window size, and have the app always start at the same stretched/resized GUI size....ie; some setting under Miscellaneous options of  'set current GUI window size as default'...

Do you know if something like that is doable in SDL2?

**Reply by midwan** [reply]
There is an open issue with this request already, https://github.com/BlitterStudio/amiberry/issues/1049 - that should cover this case as well, I think.

**Comment by giantclambake** [comment]
In #1049 I believe that's regarding the actual emulator window, not so much the GUI window ....ie; emulator window starts at 720x568 and OP resizes that window (which return to 720x568 after invoking GUI).

It is the same operation however, just applied to the GUI instead of emulator window....ie; GUI opens at 800x600, I can drag/resize that, start the emulator...but if I hit F12 to invoke the GUI it returns to 800x600 not the drag/resized value.

SO same thing in a different place -- if you consider #1049 can cover both cases, that would be grand =)

**Comment by aerno** [comment]
I just found out Amiberry can also run on X86 nowadays, excellent news ! I love the app on RPi and now even on the 'big' computer.

I'm running into the same issue, GUI and emulation windows are very small on a 4k monitor.

Emulation window can be solved by unselecting 'Correct aspect ratio'. It will show at about 3/4 of my screen :-)

As for the configuration window, it doesn't remember size at the moment. Resizing it manually works fine.

I'm testing with the Github snapshot of today BTW.

Would a 'Double Pixel Config Window' option under Miscellaneous be an idea ? 

That way the config window remains as is on RPi systems, but on modern 4k screens the window could be scaled to double the size on startup & return from emulation.

Thanks, Arno

**Comment by giantclambake** [comment]
@aerno  Just so you know, the GUI has been resized since this discussion was started, and now opens with a default size (I think?) of 800x720, and one can also change font size as well. This has gone a long way towards alleviating the concerns I had here initially -- for example, screengrab of my desktop using the native display resolution of 1920x1080 @ 60Hz ...

![ex](https://imgur.com/16Jv67V.png)

I can drag/resize the GUI window, but such is constrained by the 10:9 aspect ratio (otherwise you end up with black borders at top/bottom/sides). As said, changes made have improved the GUI to the point it really doesn't bother me anymore, but that said I don't own a 4K monitor.  

Might I ask you to attach a screenshot of your desktop with the amiberry GUI started? That'd give me/us a good idea of what you're facing =)

Wrt the latest amiberry v6.x.x preview branch, the GUI opens in it's own/separate window. That would make me wonder (albeit without knowing if it's possible or not ;), if some amiberry.conf params might be possible (holding them under the auspices of 'GUI theming' ;) ...ie; (example values)...

gui_default_height=1000
gui_default_width=900

....as the GUI is no longer constrained by the default_height/default_width params. 

The emulator window however, is a different kettle fish =)

**Comment by aerno** [comment]
Sure. I can just read the letters, but they're small ;-)

I tried the 'font size' setting in the app too in the hope the window would resize as soon as a bigger font was chosen, but that didn't happen. The font goes larger within the same space. 

![screenshot](https://github.com/BlitterStudio/amiberry/assets/36167270/47532dd2-b1c8-46dc-89de-eba43e790624)

On first start the emulation window is also very small, but as soon as I load the my saved profile and toggle 'Auto Crop' once and leave it Off the emulation window is perfect in my opinion. 27" of Amiga galore.

![screenshot2](https://github.com/BlitterStudio/amiberry/assets/36167270/7a5526f8-d275-4657-b8aa-33cd626bf564)

**Reply by midwan** [reply]
Amiberry already uses the `SDL_WINDOW_ALLOW_HIGHDPI` SDL2 flag, when creating the window. This allows it to scale according to the window manager DPI scaling option. Assuming that you have that set to something other than 100% of course!

I cannot really test this behavior however, as I don't have a 4K monitor here. I'm just following the documentation. :)

The alternative would be to do something manually, like a "x2" mode for instance.

**Reply by aerno** [reply]
The emulation window scaling seems to work fine. 

Only oddity I did notice is that when I forget to click 'Load configuration' and start the emulation directly into AROS the emulation window is small. Later when loading my saved configuration and start the emulation the window size is remembered and stays small.

When I first click Load configuration and start the emulation the emulation window is big (scaled). I did uncheck 'Correct aspect rario' in my saved configuration to get the emulation window to scale so big on my display.

I've set scaling to 150% at this moment in the Cinnemon/Gnome settings menu. 

For the settings window a 'GUI x2' mode in say the Miscellaneous menu would be great. 

That way it won't disturb RPi users when they're on 1080P or 720P and don't check the button, but high DPI users can check it to make the GUI more readable by default.

When I scale the GUI with the mouse it does scale very cool BTW, seems linear. MUI how its meant to be ;-) Every time I return from emulation its back at 1on1 again. 

I presume on RPi the SDL2 layer will use the framebuffer for its graphics output ?  If thats detectable by Amiberry you could ignore the 'GUI x2' flag when outputting to a framebuffer to prevent issues.

When I can help testing on a high DPI monitor just let me know, I can rebuild Amiberry easily anytime and test.

**Reply by giantclambake** [reply]
@aerno ~ thanks for the screenshots, it really helps elucidate the issue ;)

**Reply by midwan** [reply]
This is now implemented, with the help of a new option (`window_scaling`) in `amiberry.conf`. The wiki has been updated accordingly, as well: https://github.com/BlitterStudio/amiberry/wiki/Amiberry.conf-options

Only the `preview` and `master-gfx-rewrite` branches have it for now, not sure it's worth doing in `master` (the plan is to merge `master-gfx-rewrite` there soon anyway).

**Comment by aerno** [comment]
@midwan I switched to the 'remotes/origin/master-gfx-rewrite' branch and checked that one out.

git branch -a
* (HEAD detached at origin/master-gfx-rewrite)
  list
  master
  remotes/origin/HEAD -> origin/master
  remotes/origin/master
  remotes/origin/master-gfx-rewrite
  remotes/origin/preview
  remotes/origin/statusline

The new 'window_scaling' setting in the amiberry.conf seems to work fine. I can set the scaling to 1.5 in the configuration and the Amiberry GUI window becomes 50% larger indeed. Seems fine to me.

Idea for 'emulation_scaling'  ? To scale the emulation window on start of emulation proportional too ? 

---

I did notice a related scaling issue, but it seems to me its not linked to the new 'windows_scaling' :

For this change I tested with different Gnome desktop scaling settings (100, 150 and 200) and did notice Amiberry has a kind of reverse scaling (the GUI becomes smaller when the desktop scaling goes up) already built in.

Could it be the display DPI is already used for calculation the GUI Window size on another spot ?

It doesn't matter if I set 'window_scaling' to  1.0 or 1.5 in Amiberry, the GUI window becomes smaller when the desktop scaling goes up.

Its visible in the screenshots (with Amiberry window_scaling at 1.5) I will upload next.

**Comment by aerno** [comment]
Amiberry scaling tests:

Desktop 100% - Amiberry 150%:
![desktop100-amiberry150](https://github.com/BlitterStudio/amiberry/assets/36167270/301cd2e6-1ad8-49a3-91dd-3b31cd956666)

Desktop 100% - Amiberry 200%:
![desktop100-amiberry200](https://github.com/BlitterStudio/amiberry/assets/36167270/2ee1ee91-cc12-4fdd-925b-f8288638b448)

Desktop 150% - Amiberry 150%:
![desktop150-amiberry150](https://github.com/BlitterStudio/amiberry/assets/36167270/41dc5e48-e018-4b9a-b0ac-6e67b5b70cf2)

Desktop 150% - Amiberry 200%:
![desktop150-amiberry200](https://github.com/BlitterStudio/amiberry/assets/36167270/bea7b0aa-edd5-4eb4-aab0-11db442cb0d9)

Desktop 200% - Amiberry 150%:
![desktop200-amiberry150](https://github.com/BlitterStudio/amiberry/assets/36167270/68dde009-125e-433b-90ae-d7210f724e38)

Desktop 200% - Amiberry 200%
![desktop200-amiberry200](https://github.com/BlitterStudio/amiberry/assets/36167270/8e92c27c-f829-4ad3-a240-dc3f0195f35f)

---

Amiberry reverse scaling demonstration (desktop 100,150,200 and Amibery 100):

Desktop 100 - Amiberry 100:
![desktop100-amiberry100](https://github.com/BlitterStudio/amiberry/assets/36167270/56210560-8273-472b-a971-ccea4ed0aa94)

Desktop 150 - Amiberry 100:
![desktop150-amiberry100](https://github.com/BlitterStudio/amiberry/assets/36167270/c6dbc919-bdb0-4a3b-9e50-cfd83614fa99)

Desktop 200 - Amiberry 100:
![desktop200-amiberry100](https://github.com/BlitterStudio/amiberry/assets/36167270/84322386-f222-481d-b07e-a9d182acea97)

As shown the Amiberry GUI becomes smaller in screen size as soon as the Desktop Scaling goes up.

**Reply by midwan** [reply]
That's really weird.

There is an SDL2 option to set a window as "allowed for High DPI", but the implementation of that is back-end specific. It's supposed to work as expected on MacOS, doesn't work on Windows (which we're not using on Amiberry anyway), and I'm not sure how it behaves on Linux window managers. But it seems to be doing it backwards...

Did the same thing happen with the previous version of Amiberry as well, before the scaling option I added? Perhaps it's triggered by that somehow.

**Comment by aerno** [comment]
I will test tomorrow with version 5.7.0. 

I did notice with the git snapshot  of yesterday that the emulation window doesn't go large by default after loading the configuration. 

@midman It can be mentioned in the README.md that Amiberry 5.7.0 and later can be built and work on Rocky Linux 9.4 too. Would you like me to edit the README.md or can you do it ?

**Comment by aerno** [comment]
@midwan Version 5.7.0 does the same: It has inverse GUI Window scaling dependant on the Gnome Display->MonitorScale setting. 

Ie. when the MonitorScale value goes up from 100->150 the Amiberry GUI window becomes smaller, while it should go bigger.

It doesn't depend on the startup of Amiberry, the effect happens instantaneously when switching from 100->150 and back.

I presume its and SDL2 bug ?

**Reply by midwan** [reply]
Can you test in `preview` after getting the latest? I've removed the HighDPI setting from the window as a test, with 55cfc63321f718be5c37c5df04ff3de5c49babd5

**Comment by aerno** [comment]
I've just tested with the Preview snapshot, where the _HIGHDPI SDL parameters have been removed, but it does exactly the same as version 5.7.0 here. The AmiBerry window is larger at DesktopScaling=100 than at DesktopScaling=150.

I will do a test without OpenGL. See if its related.

**Comment by aerno** [comment]
I tried rebuilding Amiberry without OpenGL support (-DUSE_OPENGL=OFF), but I cannot seem to really disable it:

strace -f amiberry 2>&1| grep -i libgl
[pid 19123] openat(AT_FDCWD, "/opt/emu/amiga/amiberry/lib/libGL.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
[pid 19123] openat(AT_FDCWD, "/opt/multimedia/libs/lib/libGL.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
[pid 19123] openat(AT_FDCWD, "/lib64/libGL.so.1", O_RDONLY|O_CLOEXEC) = 14
[pid 19123] openat(AT_FDCWD, "/opt/emu/amiga/amiberry/lib/libGLX.so.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
[pid 19123] openat(AT_FDCWD, "/opt/multimedia/libs/lib/libGLX.so.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
[pid 19123] openat(AT_FDCWD, "/lib64/libGLX.so.0", O_RDONLY|O_CLOEXEC) = 14
[pid 19123] openat(AT_FDCWD, "/opt/emu/amiga/amiberry/lib/libGLdispatch.so.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
[pid 19123] openat(AT_FDCWD, "/opt/multimedia/libs/lib/libGLdispatch.so.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
[pid 19123] openat(AT_FDCWD, "/lib64/libGLdispatch.so.0", O_RDONLY|O_CLOEXEC) = 14
[pid 19123] openat(AT_FDCWD, "/opt/emu/amiga/amiberry/lib/libGLX_nvidia.so.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
[pid 19123] openat(AT_FDCWD, "/opt/multimedia/libs/lib/libGLX_nvidia.so.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
[pid 19123] openat(AT_FDCWD, "/lib64/libGLX_nvidia.so.0", O_RDONLY|O_CLOEXEC) = 14
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGL.so.1.7.0", O_RDONLY) = 14
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLX.so.0.0.0", O_RDONLY) = 14
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLdispatch.so.0.0.0", O_RDONLY) = 14
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLX_nvidia.so.550.78", O_RDONLY) = 14
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLX_nvidia.so.550.78", O_RDONLY) = 14
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLX_nvidia.so.550.78", O_RDONLY) = 15
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLX_nvidia.so.550.78", O_RDONLY) = 14
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLX_nvidia.so.550.78", O_RDONLY) = 15
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGL.so.1.7.0", O_RDONLY) = 14
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGL.so.1.7.0", O_RDONLY) = 15
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGL.so.1.7.0", O_RDONLY) = 14
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGL.so.1.7.0", O_RDONLY) = 15
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGL.so.1.7.0", O_RDONLY) = 16
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLX.so.0.0.0", O_RDONLY) = 16
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLdispatch.so.0.0.0", O_RDONLY) = 16
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLX_nvidia.so.550.78", O_RDONLY) = 16
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGL.so.1.7.0", O_RDONLY) = 16
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLX.so.0.0.0", O_RDONLY) = 16
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLdispatch.so.0.0.0", O_RDONLY) = 16
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLX_nvidia.so.550.78", O_RDONLY) = 16
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGL.so.1.7.0", O_RDONLY) = 16
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLX.so.0.0.0", O_RDONLY) = 16
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLdispatch.so.0.0.0", O_RDONLY) = 16
[pid 19123] openat(AT_FDCWD, "/usr/lib64/libGLX_nvidia.so.550.78", O_RDONLY) = 16

I guess that test isn't possible for me at the moment. 

Maybe OpenGL explains the smooth scaling of the AmiBerry GUI.

**Reply by midwan** [reply]
Amiberry does not use OpenGL by default. You can enable it manually, but that back-end is still WIP and doesn't quite work as it should (I'm trying to implement a CRT shader there).
By default, it lets SDL2 choose the best available back-end itself. I'm not sure what's wrong with the scaling issue, I think I've followed what's mentioned in the documentation but perhaps I'm missing something...

**Comment by StusBrainDump** [comment]
Sorry to jump into this but I think it's within context .... is there an option to make the main GUI always start in fullscreen in macOS?

**Reply by midwan** [reply]
Yes, there is an option about this in `amiberry.conf` that controls the default screen mode (Windowed, Full-WIndow, Fullscreen). https://github.com/BlitterStudio/amiberry/wiki/Amiberry.conf-options
